### PR TITLE
feat: add dev build

### PR DESCRIPTION
**Pull request by mabaasit**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: .evergreen/functions.yml [hunk 1]
```diff
@@ -301,7 +301,12 @@ functions:
           set -e
           # Load environment variables
           eval $(.evergreen/print-compass-env.sh)
-
+          # The artifact file names are generated (in target.js) and added to the expansions
+          if [[ "${project}" == "10gen-compass-main" && "${branch_name}" == "main"  ]]; then
+            # In order to correclty use dev version identifer based on datetime, we will use
+            # evergreen's created_at expansion which is formatted as: 24_05_16_14_52_37.
+            export DEV_VERSION_IDENTIFIER=$(awk '{gsub("_", ""); print}' <<< ${created_at})
```
Comment (mcasimir) [inline .evergreen/functions.yml:308]: NIT: should we put this code inside `.evergreen/print-compass-env.sh`?

File: .evergreen/functions.yml [hunk 2]
```diff
@@ -448,6 +457,54 @@ functions:
...
+            # Check if the current version is greater than the version we are trying to publish.
+            if [[ $VERSION_ID -le $CURRENT_VERSION_ID ]]; then
+              echo "Published version ${CURRENT_VERSION_ID} is greater than version ${VERSION_ID} trying to publish. Exiting..."
+              exit 1
+            fi
+          fi
+
+          VERSION="v0.0.0-dev.$VERSION_ID"
+          PUBLISHED_AT=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
+          echo "{\"id\": $VERSION_ID, \"version\": \"v0.0.0-dev.$VERSION_ID\", \"published_at\": \"$PUBLISHED_AT\", \"draft\": true}" > $FILE_NAME
```
Comment (mcasimir) [inline .evergreen/functions.yml:497]: do we need quotes to surround the version?

```suggestion
          echo "{\"id\": \"$VERSION_ID\", \"version\": \"v0.0.0-dev.$VERSION_ID\", \"published_at\": \"$PUBLISHED_AT\", \"draft\": true}" > $FILE_NAME
```

do we need a json here? how we would use these info, if they are not useful (especially the draft: true), we could rather have https://mciuploads.s3.amazonaws.com/${project}/compass/dev/LATEST and the content is just the version

File: .evergreen/functions.yml (hunk 1 above)
Comment (mcasimir) [reply .evergreen/functions.yml:308]: Also NIT: does it make sense to export the entire `DEV_VERSION=0.0.0-${somedatetime}` if we need to concatenate everywhere

File: .evergreen/functions.yml (hunk 2 above)
Comment (mcasimir) [inline .evergreen/functions.yml:497]: what is more comfortable for semver comparison "v0.0.0-dev.$VERSION_ID" or just "0.0.0-dev.$VERSION_ID"

File: packages/hadron-build/lib/validate-build-config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+
+const platformRequiredKeys = {
+  darwin: ['dmg_background', 'app_category_type', 'icon'],
+  linux: ['deb_section', 'rpm_categories', 'icon'],
+  win32: ['icon', 'favicon_url', 'loading_gif', 'background', 'banner']
+};
+
+const validateBuildConfig = (
```
Comment (mcasimir) [inline packages/hadron-build/lib/validate-build-config.js:8]: ❤️

File: .evergreen/functions.yml (hunk 2 above)
Comment (mabaasit) [reply .evergreen/functions.yml:497]: i dont have a strong preference here, but i wanted to keep the published datetime in the json. we can totally drop everything and just have the versionId in the file - `0.0.0-dev.YYDDMMHHMMSS` format and then use that in update api.

Comment (mabaasit) [reply .evergreen/functions.yml:497]: it works with both.

File: .evergreen/functions.yml (hunk 1 above)
Comment (mabaasit) [reply .evergreen/functions.yml:308]: Yeah, i'd add it there and make concatenation part of it as well.

File: .evergreen/publish-dev-release.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+URL="https://mciuploads.s3.amazonaws.com/${EVERGREEN_PROJECT}/compass/dev/$1"
+TMP_FILE=$(mktemp)
+CURRENT_VERSION="0.0.0-dev.0"
+
+if [ "$(curl -s -w "%{http_code}" -o "$TMP_FILE" "$URL" | tail -n1)" -eq 200 ]; then
+  CURRENT_VERSION=$(xargs < "$TMP_FILE")
+fi
+
+npm install -g semver
```
Comment (mcasimir) [inline .evergreen/publish-dev-release.sh:24]: this shouldn't be needed with `npx`

File: .evergreen/publish-dev-release.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  exit 0;
+fi
+
+if [[ "${EVERGREEN_BRANCH_NAME}" != "main" ]]; then
+  echo "Trying to publish main compass (dev build) from ${EVERGREEN_BRANCH_NAME} branch. Skipping...";
+  exit 0;
+fi
+
+URL="https://mciuploads.s3.amazonaws.com/${EVERGREEN_PROJECT}/compass/dev/$1"
+TMP_FILE=$(mktemp)
```
[the remark this answers is not in the document]
Comment (mabaasit) [reply .evergreen/publish-dev-release.sh]: cleaned in 9041406

**Review by mcasimir — APPROVED** [review]

**Pull request merged**